AAFaceDll/config.txt
;VERBOSITY sets the verbosity of the logger. possible modes:
;	SPAM
;	INFO
;	WARNING
;	ERROR
;	CRITICAL_ERROR

VERBOSITY SPAM

;DISABLE disables certain features:
;	GENERAL				: disables hotkeys
;	FACE				: disables face dropdowns
;	HAIR				: disables hair stuff
;	HAIR_SKIPINVALID	: disables the thing where the hair keeps scrolling on invalid slots.
;							note that this feature is based on the info given by the list files,
;							and might therefor view hairs as existing when they are in fact not,
;							and apparently also as missing when they are not
;	FACEDETAILS			: disables facedetail edit boxes
;	BODY_COLOR			: disables body color edit boxes
;	LIMITS				: disables the unlimited sliders that weren't changed by the unlocked editor
;	PRE_BACKUP			: disables automatic backup of the current card when saving
;	POST_BACKUP			: disables automatic backup of the new card, before the old cardface was reapplied
;	BACKUP_UNIQUE		: on default, backups will be seeded with a timestamp to assure that they are unique.
;							if this is disabled, there will only be up to 2 backups per card (pre, post)

DISABLE HAIR_SKIPINVALID
DISABLE POST_BACKUP
DISABLE BACKUP_UNIQUE

; the minimum and maximum zoom can be changed using ZOOM [MIN/MAX] float
; MIN means zooming out; default value is 1.5.
; MAX means zooming in; default value is 0.1. smaller = closer.

;Set Zoom Properties
ZOOM MAX 0.01 ; default is 0.1

; virtual key codes: https://msdn.microsoft.com/en-us/library/windows/desktop/dd375731%28v=vs.85%29.aspx
; ascii key codes (1-127): https://msdn.microsoft.com/en-us/library/60ecse8t%28v=vs.80%29.aspx
; ascii key codes (128-255) (might not work/differ depending on your keyboard): 
;						https://msdn.microsoft.com/en-us/library/9hxt0028%28v=vs.80%29.aspx

; hotkeydefinitions start with KEY. Available options:
; MODs: ALT, CTRL, SHIFT 	: modifier keys that are part of the hotkey. multiple can be used.
; KEYVAL:
;	VKEY:				: following is a virtual key id in hexadecimal (this is default)
;	ASCII:				: following is an ascii value in hexadecimal
;		<param>			: the actual key code. must be followed and given in hexadecimal as parsed by strtol(,,16)
; CONTEXTs:
;    ACTIVE:			: window must exist and currently be active. no effect on DIALOG_ALL, ALL or PREVIEW_WINDOW
;    FOCUS:				: window is active and it or one of its child must have current focus. 
;							for dialogs, the upper bar might also be in focus if the given dialog is active
;						: these values are optional; if not given, it defaults to FOCUS
;		DIALOG_SYSTEM
;		DIALOG_FIGURE
;		DIALOG_CHEST
;		DIALOG_BODY_COLOR
;		DIALOG_FACE
;		DIALOG_EYES
;		DIALOG_EYE_COLOR
;		DIALOG_FACE_DETAILS
;		DIALOG_HAIR			: only valid while this dialog is focused
;		DIALOG_HAIR_COLOR
;		DIALOG_CHARACTER
;		DIALOG_PERSONALITY
;		DIALOG_TRAITS	
;	    DIALOG_ALL		: valid while any dialog is on focus
;	    PREVIEW_WINDOW	: valid while the character-rendering window is focused
;	    ALL (default)	: valid anywhere, everywhere
; FUNCTIONs: HAIR_FLIP		: flips the currently selected hair
;	     HAIR_ADD		: adds iparam to the current hairs index
;	     HAIR_SET		: sets current hair to iparam
;	     HAIR_SIZEADD	: adds iparam to the current hairs size (slider)
;	     HAIR_SIZESET	: sets value of size slider to iparam
;		 FACEDETAILS_ADDGLASSES : adds iparam to current glasses value
;	     FACEDETAILS_SETGLASSES : sets current glasses to iparam
;		 BODYCOLOR_ADDTAN		: adds iparam to current tans value
;		 BODYCOLOR_SETTAN		: sets current tan to iparam
; 	     ZOOM_ADD		: zooms in or out, depending on float-param
; 	     ZOOM_SET		: sets zoom to float parameter
;	     TILT_ADD		: tilts the view by fparam (values from 0 to 6.28)
;	     TILT_SET		: sets tilt to fparam (values from 0 to 6.28)
;		 POSE_ADD		: add iparam to the current pose
;		 POSE_APPLY		: apply current pose, as if the apply button was clicked
;	     NOP			: does nothing. originally supposed to be able to overwrite
;							default keys, but i ended up using accels for easy hold-key-behavour,
;							so that doesnt work now
; PARAM:     INT		: 32 bit integer parameter. must be followed as decimal number as parsed by atoi
;			 FLOAT		: 32 bit float parameter. must be followed as float number as parsed by atof

; these parameters can the chained together in no particular order. except for MOD (and PARAM), only one may be used. example:
;KEY MOD:ALT MOD:CTRL KEYVAL:50 FUNCTION:HAIR_FLIP	; fine
;KEY KEYVAL:50 FUNCTION:HAIR_FLIP FUNCTION:ZOOM_ADD   ; not fine, 2 functions
;KEY FUNCTION:HAIR_FLIP								; which key, dumbass
; if you want a key to do multiple things, maybe have different meaning in different contexts, define the same key
; with the same mod multiple times in succession (with no other key definition inbetween). example:
;KEY KEYVAL:4C CONTEXT:FOCUS:DIALOG_HAIR FUNCTION:HAIR_ADD PARAM:INT:1
;KEY KEYVAL:4C CONTEXT:FOCUS:DIALOG_FACE_DETAILS FUNCTION:FACEDETAILS_ADD PARAM:INT:1
;makes L choose the next hair if hair dialog is in focus, or next glasses if face details dialog is in focus 
; note that it will probably not warn you about invalid keys cause im lazy.

; testing different kinds of contexts
KEY CONTEXT:ALL KEYVAL:50 FUNCTION:HAIR_FLIP ; P
KEY CONTEXT:DIALOG_HAIR KEYVAL:50 FUNCTION:HAIR_FLIP ; P
KEY CONTEXT:DIALOG_ALL KEYVAL:4F FUNCTION:HAIR_FLIP ;	O
KEY CONTEXT:PREVIEW_WINDOW KEYVAL:49 FUNCTION:HAIR_FLIP ; I

;prev and next hair test
KEY CONTEXT:ACTIVE:DIALOG_HAIR KEYVAL:4C FUNCTION:HAIR_ADD PARAM:INT:1; L
KEY CONTEXT:ACTIVE:DIALOG_HAIR KEYVAL:4B FUNCTION:HAIR_ADD PARAM:INT:-1; K

;hair size test
KEY CONTEXT:ACTIVE:DIALOG_HAIR KEYVAL:4C MOD:CTRL FUNCTION:HAIR_SIZEADD PARAM:INT:1
KEY CONTEXT:ACTIVE:DIALOG_HAIR KEYVAL:4B MOD:CTRL FUNCTION:HAIR_SIZEADD PARAM:INT:-1

;ascii test
KEY CONTEXT:ALL KEYVAL:ASCII:3D FUNCTION:HAIR_FLIP

;pose scroll keys
KEY CONTEXT:ACTIVE:DIALOG_SYSTEM KEYVAL:57 MOD:SHIFT FUNCTION:POSE_ADD PARAM:INT:1	;W
KEY CONTEXT:ACTIVE:DIALOG_SYSTEM KEYVAL:57 MOD:SHIFT FUNCTION:POSE_APPLY

KEY CONTEXT:ACTIVE:DIALOG_SYSTEM KEYVAL:53 MOD:SHIFT FUNCTION:POSE_ADD PARAM:INT:-1	;S
KEY CONTEXT:ACTIVE:DIALOG_SYSTEM KEYVAL:53 MOD:SHIFT FUNCTION:POSE_APPLY

AAFaceDll/Config.cpp
#include "Config.h"
#include "Logger.h"
#include <Windows.h>

Config g_config;

const Config::Keywords Config::Commands[] = {
	{ "KEY", &Config::InterpretKey },
	{ "ZOOM", &Config::InterpretZoom },
	{ "DISABLE", &Config::InterpretDisable },
	{ "VERBOSITY", &Config::InterpretLogger }

};

Config::Config() : zoomMin(1.5f),zoomMax(0.01f),
				disableFlags(0),
				logPrio(Logger::Priority::WARN)
{
}

//KEY ALT CTRL SHIFT 1
Config::Config(const char* file) : Config()
{
	HANDLE f = CreateFile(file,GENERIC_READ,FILE_SHARE_READ,NULL,OPEN_EXISTING,0,NULL);
	if (f == NULL || f == INVALID_HANDLE_VALUE) {
		LOGPRIO(Logger::Priority::WARN) << "Could not open config file";
		return; //failed to open
	}
	DWORD fhi; //this really shouldnt be needed, but GetFilieSize crashes if its not given (it shouldnt, but it does)
	DWORD flo = GetFileSize(f,&fhi);
	char* buffer = new char[flo+1];
	buffer[flo] = '\0';
	DWORD bufferLength = 0;
	ReadFile(f,buffer,flo,&bufferLength,NULL);
	CloseHandle(f);
	if (bufferLength == 0) return; //nothing read for some reason
	buffer[bufferLength] = '\0';

	//read buffer, line for line
	char* it = NULL,*nextLine = buffer;
	do {
		it = nextLine;
		GetLine(it,&nextLine);
		//get first token
		char* token;

		token = GetToken(it,&it);
		if (token != NULL) {
			//possible keywords:
			bool found = false;
			for (int i = 0; i < sizeof(Commands) / sizeof(Commands[0]); i++) {
				if (strcmp(token,Commands[i].key) == 0) {
					(this->*(Commands[i].handler))(it);
					found = true;
					break;
				}
			}
			if(!found) {
				LOGPRIO(Logger::Priority::WARN) << "Unkown Config Command \"" << token << "\" found in config file\n";
			}
		}

	} while (nextLine != NULL);
	delete[] buffer;
}


Config::~Config()
{
}

bool Config::InterpretKey(char * str)
{
	char* nextToken = str;
	char* token = NULL;
	Hotkey key;
	//i really should have made this a data structure instead. fuck.
	while (nextToken != NULL && (token = GetToken(nextToken,&nextToken))) {
		if (StartsWith(token,"MOD:")) {
			token += 4;
			if (strcmp(token,"ALT") == 0) {
				key.alt = true;
			}
			else if (strcmp(token,"CTRL") == 0) {
				key.ctrl = true;
			}
			else if (strcmp(token,"SHIFT") == 0) {
				key.shift = true;
			}
			else {
				LOGPRIO(Logger::Priority::WARN) << "Unknown Modification \"" << token << "\" in config file\n";
			}
		}
		else if (StartsWith(token,"CONTEXT:")) {
			token += 8;
			if(StartsWith(token, "FOCUS:")) {
				token += 6;
				key.contextKind = Hotkey::FOCUS;
			}
			else if(StartsWith(token, "ACTIVE:")) {
				token += 7;
				key.contextKind = Hotkey::ACTIVE;
			}
			else {
				key.contextKind = Hotkey::FOCUS;
				//gonna allow it, assuming they know what they do
				//g_Logger << Logger::Priority::WARN << "Unrecognized context kind in " << token << "\n";
			}
			if (StartsWith(token, "DIALOG_")) {
				token += 7;
				if(strcmp(token, "SYSTEM") == 0) {
					key.context = Hotkey::DIALOG_SYSTEM;
				}
				else if(strcmp(token,"FIGURE") == 0) {
					key.context = Hotkey::DIALOG_FIGURE;
				}
				else if (strcmp(token,"CHEST") == 0) {
					key.context = Hotkey::DIALOG_CHEST;
				}
				else if (strcmp(token,"BODY_COLOR") == 0) {
					key.context = Hotkey::DIALOG_BODY_COLOR;
				}
				else if (strcmp(token,"FACE") == 0) {
					key.context = Hotkey::DIALOG_FACE;
				}
				else if (strcmp(token,"EYES") == 0) {
					key.context = Hotkey::DIALOG_EYES;
				}
				else if (strcmp(token,"EYE_COLOR") == 0) {
					key.context = Hotkey::DIALOG_EYE_COLOR;
				}
				else if(strcmp(token, "EYEBROWS") == 0) {
					key.context = Hotkey::DIALOG_EYEBROWS;
				}
				else if (strcmp(token,"FACE_DETAILS") == 0) {
					key.context = Hotkey::DIALOG_FACE_DETAILS;
				}
				else if (strcmp(token,"HAIR") == 0) {
					key.context = Hotkey::DIALOG_HAIR;
				}
				else if (strcmp(token,"HAIR_COLOR") == 0) {
					key.context = Hotkey::DIALOG_HAIR_COLOR;
				}
				else if (strcmp(token,"CHARACTER") == 0) {
					key.context = Hotkey::DIALOG_CHARACTER;
				}
				else if (strcmp(token,"PERSONALITY") == 0) {
					key.context = Hotkey::DIALOG_PERSONALITY;
				}
				else if (strcmp(token,"TRAITS") == 0) {
					key.context = Hotkey::DIALOG_TRAITS;
				}
				else if (strcmp(token,"ALL") == 0) {
					key.context = Hotkey::DIALOG_ALL;
				}
				else {
					LOGPRIO(Logger::Priority::WARN) << "Unknown Dialog \"" << token << "\" in config file\n";
				}
			}
			else if (strcmp(token,"PREVIEW_WINDOW") == 0) {
				key.context = Hotkey::PREVIEW_WINDOW;
			}
			else if (strcmp(token,"ALL") == 0) {
				key.context = Hotkey::ALL;
			}
			else {
				//unrecognized context
				LOGPRIO(Logger::Priority::WARN) << "Unknown Context \"" << token << "\" in config file\n";
			}
		}
		else if (StartsWith(token,"KEYVAL:")) {
			token += 7;
			if(StartsWith(token, "VIRT:")) {
				token += 5;
				key.isVirt = true;
			}
			else if(StartsWith(token, "ASCII:")) {
				token += 6;
				key.isVirt = false;
			}
			else {
				key.isVirt = true;
			}
			key.key = strtol(token,NULL,16);
		}
		else if (StartsWith(token,"FUNCTION:")) {
			token += 9;
			if (strcmp(token,"HAIR_FLIP") == 0) {
				key.func = Hotkey::HAIR_FLIP;
			}
			else if (strcmp(token,"HAIR_ADD") == 0) {
				key.func = Hotkey::HAIR_ADD;
			}
			else if (strcmp(token,"HAIR_SET") == 0) {
				key.func = Hotkey::HAIR_SET;
			}
			else if (strcmp(token,"HAIR_SIZEADD") == 0) {
				key.func = Hotkey::HAIR_SIZEADD;
			}
			else if (strcmp(token,"HAIR_SIZESET") == 0) {
				key.func = Hotkey::HAIR_SIZESET;
			}
			else if (strcmp(token,"FACEDETAILS_ADDGLASSES") == 0) {
				key.func = Hotkey::FACEDETAILS_ADDGLASSES;
			}
			else if (strcmp(token,"FACEDETAILS_SETGLASSES") == 0) {
				key.func = Hotkey::FACEDETAILS_SETGLASSES;
			}
			else if (strcmp(token,"BODYCOLOR_ADDTAN") == 0) {
				key.func = Hotkey::BODYCOLOR_ADDTAN;
			}
			else if (strcmp(token,"BODYCOLOR_SETTANS") == 0) {
				key.func = Hotkey::BODYCOLOR_SETTAN;
			}
			else if(strcmp(token, "ZOOM_ADD") == 0) {
				key.func = Hotkey::ZOOM_ADD;
			}
			else if (strcmp(token,"ZOOM_SET") == 0) {
				key.func = Hotkey::ZOOM_SET;
			}
			else if (strcmp(token,"TILT_ADD") == 0) {
				key.func = Hotkey::TILT_ADD;
			}
			else if (strcmp(token,"TILT_SET") == 0) {
				key.func = Hotkey::TILT_SET;
			}
			else if (strcmp(token,"POSE_ADD") == 0) {
				key.func = Hotkey::POSE_ADD;
			}
			else if (strcmp(token,"POSE_APPLY") == 0) {
				key.func = Hotkey::POSE_APPLY;
			}
			else if(strcmp(token, "NOP") == 0) {
				key.func = Hotkey::NOP;
			}
			else {
				LOGPRIO(Logger::Priority::WARN) << "Unknown Function \"" << token << "\" in config file\n";
			}
		}
		else if(StartsWith(token, "PARAM:")) {
			token += 6;
			if(StartsWith(token, "INT:")) {
				token += 4;
				key.iparam = atoi(token);
			}
			else if(StartsWith(token, "FLOAT:")) {
				token += 6;
				key.fparam = (float)atof(token);
			}
			else {
				LOGPRIO(Logger::Priority::WARN) << "Unknown Param \"" << token << "\" in config file\n";
			}
		}
		else {
			//unrecognized parameter type
			LOGPRIO(Logger::Priority::WARN) << "Unknown Parameter Type \"" << token << "\" in config file\n";
		}
	}
	if (key.key == -1) {
		//no virtual key id given
		LOGPRIO(Logger::Priority::WARN) << "Key Definition without KEYVAL found\n";
		return false;
	}
	else if(key.key <= 0 || key.key > 255) {
		LOGPRIO(Logger::Priority::WARN) << "key with invalid KEYVAL " << key.key << " found.\n";
		return false;
	}
	hotkeys.push_back(key);
	return true;
}

bool Config::InterpretZoom(char* str)
{
	char* number;
	char* minmax;
	minmax = GetToken(str,&number);
	if (minmax == NULL) return false;
	number = GetToken(number,NULL);
	if (number == NULL) return false;
	float f = strtof(number,NULL);
	if (f == 0) return false;
	if (strcmp(minmax,"MAX") == 0) {
		zoomMax = f;
	}
	else if (strcmp(minmax,"MIN") == 0) {
		zoomMin = f;
	}
	else {
		return false;
	}
	return true;
}

bool Config::InterpretDisable(char * str)
{
	char* token = GetToken(str,NULL);
	if(token != NULL) {
		if(strcmp(token, "GENERAL") == 0) {
			disableFlags |= DISABLE_GENERAL;
		}
		else if(strcmp(token, "FACE") == 0) {
			disableFlags |= DISABLE_FACE;
		}
		else if (strcmp(token,"HAIR") == 0) {
			disableFlags |= DISABLE_HAIR;
		}
		else if (strcmp(token,"HAIR_SKIPINVALID") == 0) {
			disableFlags |= DISABLE_HAIR_SKIPINVALID;
		}
		else if (strcmp(token,"FACEDETAILS") == 0) {
			disableFlags |= DISABLE_FACEDETAILS;
		}
		else if (strcmp(token, "BODY_COLOR") == 0) {
			disableFlags |= DISABLE_BODY_COLOR;
		}
		else if(strcmp(token, "LIMITS") == 0) {
			disableFlags |= DISABLE_LIMITS;
		}
		else if (strcmp(token,"PRE_BACKUP") == 0) {
			disableFlags |= DISABLE_PRE_BACKUP;
		}
		else if (strcmp(token,"POST_BACKUP") == 0) {
			disableFlags |= DISABLE_POST_BACKUP;
		}
		else if (strcmp(token,"BACKUP_UNIQUE") == 0) {
			disableFlags |= DISABLE_BACKUP_UNIQUE;
		}
		else {
			LOGPRIO(Logger::Priority::WARN) << "Unknown Hook " << token << " disabled in config\n";
		}
	}
	return false;
}

bool Config::InterpretLogger(char* str) {
	char* token = GetToken(str,NULL);
	if(token != NULL) {
		if(strcmp(token, "SPAM") == 0) {
			logPrio = Logger::Priority::SPAM;
			g_Logger.SetPriority(logPrio);
		}
		else if(strcmp(token, "INFO") == 0) {
			logPrio = Logger::Priority::INFO;
			g_Logger.SetPriority(logPrio);
		}
		else if (strcmp(token,"WARNING") == 0) {
			logPrio = Logger::Priority::WARN;
			g_Logger.SetPriority(logPrio);
		}
		else if (strcmp(token,"ERROR") == 0) {
			logPrio = Logger::Priority::ERR;
			g_Logger.SetPriority(logPrio);
		}
		else if (strcmp(token,"CRITICAL_ERROR") == 0) {
			logPrio = Logger::Priority::CRIT_ERR;
			g_Logger.SetPriority(logPrio);
		}
		else {
			LOGPRIO(Logger::Priority::WARN) << "Invalid logger verbosity " << token << " in config file\n";
			return false;
		}
	}
	else {
		return false;
	}
	return true;
}

/*
* Checks if the string at str starts with word
*/
bool Config::StartsWith(const char * str,const char * word)
{
	while (*word && *str && (*str++ == *word++));
	if (*word == '\0') return true;
	return false;
}

const std::vector<Config::Hotkey>& Config::GetHotkeys() const
{
	return hotkeys;
}

float Config::GetZoomMin() const
{
	return zoomMin;
}
float Config::GetZoomMax() const
{
	return zoomMax;
}

bool Config::IsDisabled(Disable check) const {
	return disableFlags & check;
}

/*
* Modifies str so that it points to a single line, comments (;) or newlines filtered out.
* if nextLine != NULL, *nextLine becomes the start of the next line, or NULL if end was reached
*/
void Config::GetLine(char* str,char** nextLine) {
	char* it = str;
	//iterate to either ; (comment), \r\n (newline), \n (newline - unix style), \r (newline - mac style),
	//so probably just \r or \n, or \0
	while (true) {
		if (*it == ';') {
			*it++ = '\0';
			//skip rest of line
			while (*it && *it != '\r' && *it != '\n') it++;
			while (*it && *it == '\r' || *it == '\n') it++;
			if (nextLine != NULL) {
				if (*it)
					*nextLine = it;
				else
					*nextLine = NULL;
			}
			break;
		}
		else if (*it == '\r' || *it == '\n') {
			*it++ = '\0';
			while (*it && *it == '\r' || *it == '\n') it++;
			if (nextLine != NULL) {
				if (*it)
					*nextLine = it;
				else
					*nextLine = NULL;
			}
			break;
		}
		else if (*it == '\0') {
			if (nextLine != NULL) *nextLine = NULL;
			break;
		}
		it++;
	}
}

/*
* Gets next whitespace seperated token from str, nullterminates it,
* sets nextPart (if != NULL) to point to after the \0 and then
* returns a pointer to the beginning of the token (skipping whitespaces),
* or NULL if no token was found (only whitespaces and \0)
*/
char* Config::GetToken(char* str,char** nextPart) {
	while (isspace(*str) && *str != '\0') str++;
	if (*str == '\0') {
		if (nextPart != NULL) *nextPart = NULL;
		return NULL;
	}
	char* ret = str;
	while (!isspace(*str) && *str != '\0') str++;
	if (*str == '\0') {
		if (nextPart != NULL) *nextPart = NULL;
		return ret;
	}
	*str = '\0';
	if (nextPart != NULL) {
		if (*(str+1) == '\0') {
			*nextPart = NULL;
		}
		else {
			*nextPart = str+1;
		}
	}
	return ret;
}